# **Project 1 - Sales Analysis - ETL**

## Objectives

- Extract data from provided CSV files
- Clean it 
- Apply feature engineering if necessary 
- Remove any unnecessary columns
- Save to a new CSV file


## Inputs

CSV files provided:

stores data-set.csv
sales data-set.csv
features data set.csv

Renamed files to a uniform naming convention:

Sales_Features_DataSet.csv
Sales_DataSet.csv
Sales_Stores_DataSet.csv

Note: original files are stored in Data/OriginalFiles


## Outputs

CSV file created from ETL etc stored in Data/CleanedDataSets:

Sales_DataSet_Cleaned.csv


## Additional Comments

Used various AI tools to help with the ETL process:
- ChatGPT
- GitHub Copilot. 

See Documents/What_AI_Used_For.md for more details.

## Initalise Working Environment

In [ ]:
#import libraries
import os
import numpy as np
import pandas as pd

#below solution provided by chatGPT
from assets.python_files.modGlobal import modGlobal
from assets.python_files.modETL_Library import modETL
#end solution provided by chatGPT




# Section 1 - Initalisation

## Intitialise All Variables To Be Used In Global Stack 

In [ ]:
#DataFrame vars for ETL
dfSales_DataSet = None
dfSales_DataSet_Work = None
dfCleaned = None
dfCheck = None

#missing values check vars
intMissing = 0
intMissingPercent = 0
intLessThanZero = 0
dfCheck = None

#duplicates check vars
intDuplicates = 0

#stores current directory
strCurrentDir = ""

#constant for project directory
CNST_STR_PROJECT_DIR = "Project1-SalesAnalysis"

#file path constants
CNST_STR_CLEANFILEPATH = "/Data/CleanedDataSets/Sales_DataSet_Cleaned.csv"
CNST_STR_SALESFILEPATH = "/Data/Sales_DataSet.csv"

#other vars
objColumns = None

## Set Current Directory To Base Project Directory

In [3]:
#get project directory - default is jupyter notebook sub folder as that is where this file is located!
#so move back one to the project root path
# Source - https://stackoverflow.com/a/17726833
# Posted by chimpsarehungry
# Retrieved 2026-07-05, License - CC BY-SA 3.0

#get current folder
strCurrentDir = os.getcwd()

#is the last part of the path the project directory?
if not strCurrentDir.endswith(CNST_STR_PROJECT_DIR):
   #get current working directory and move back one to the project root path
   strCurrentDir =  os.path.normpath(os.getcwd() + os.sep + os.pardir)
   os.chdir(os.path.dirname(strCurrentDir))
   #change directory
   os.chdir(strCurrentDir)

#confirm current directory is project directory
print(f"Current Directory: \n {os.getcwd()}")

Current Directory: 
 /Users/rogerwilliams/Projects/Python/CourseProjects/Project1-SalesAnalysis


# Section 1 - Extraction

- Read csv files
- Get schema info -> column and row numbers
- Get first 5 records
- Get list of column datatypes
- Get detailed schema infomation

## Read csv Files Into Variable For Processing

In [4]:
#read csv file into DataFrame
dfSales_DataSet = pd.read_csv(strCurrentDir + CNST_STR_SALESFILEPATH)


## Get Schema Info - Sales_DataSet

In [5]:
print(f"Schema Information \n Rows: {dfSales_DataSet.shape[0]} Columns: {dfSales_DataSet.shape[1]}")
dfSales_DataSet.head()

Schema Information 
 Rows: 421570 Columns: 5


,Store,Dept,Date,Weekly_Sales,IsHoliday
0,1,1,05/02/2010,24924.50,False
1,1,1,12/02/2010,46039.49,True
2,1,1,19/02/2010,41595.55,False
3,1,1,26/02/2010,19403.54,False
4,1,1,05/03/2010,21827.90,False


## Observations - Sales_DataSet
421,570 rows
5 columns

Columns:

Store, Dept, Date, Weekly_Sales, IsHoliday

In [6]:
#get column data types
print (f"Data Types: \n{dfSales_DataSet.dtypes}")

Data Types: 
Store             int64
Dept              int64
Date             object
Weekly_Sales    float64
IsHoliday          bool
dtype: object


## Observations - Sales_DataSet

Date is a string - needs conversion to datetime
Other column types look ok

In [7]:
#get schema infomation
print(f"More Information:\n {dfSales_DataSet.info()}")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 421570 entries, 0 to 421569
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Store         421570 non-null  int64  
 1   Dept          421570 non-null  int64  
 2   Date          421570 non-null  object 
 3   Weekly_Sales  421570 non-null  float64
 4   IsHoliday     421570 non-null  bool   
dtypes: bool(1), float64(1), int64(2), object(1)
memory usage: 13.3+ MB
More Information:
 None


## Observations - Sales_DataSet

No missing values
File size: 13.3MB

---

# Section 2 - Transformation

- Get missing values percent
- Get duplicates (if any)


In [8]:
#get total missing values
intMissing = dfSales_DataSet.isnull().sum()

#calculate percent missing
intMissingPercent = ( intMissing / len(dfSales_DataSet) * 100)
dfCheck = pd.DataFrame( {"Missing Values": intMissing, "%" : intMissingPercent})

#show results
if dfCheck["Missing Values"].sum() !=0:
   print( dfCheck[dfCheck["Missing Values"] > 0])
else:
   print( "No Missing Values")     



No Missing Values


## Observations - Sales_DataSet

No missing values!

## Find Duplicates - Sales_DataSet

In [9]:
#get number of duplicates
intDuplicates = dfSales_DataSet.isnull().sum()
intDuplicates


Store           0
Dept            0
Date            0
Weekly_Sales    0
IsHoliday       0
dtype: int64

## Observations - Sales_DataSet

No duplicates!

## Get Unique Values - Sales_DataSet

In [10]:
#show unique values count
dfSales_DataSet.nunique()

Store               45
Dept                81
Date               143
Weekly_Sales    359464
IsHoliday            2
dtype: int64

## Observations - Sales_DataSet

Store has 45 unique values
Dept has 81 unique values
Date has 143 unique values
Weekly_Sales has 349,464 unique values
IsHoliday has 2 unique values

## Get Categorical Value Distribution

In [11]:
#get categorical distribution for object columns (non numeric)
for objColumn in ["Date"]:
    print( f"[{objColumn}] Value Distribution:")
    print( dfSales_DataSet[objColumn].value_counts())

[Date] Value Distribution:
Date
23/12/2011    3027
25/11/2011    3021
16/12/2011    3013
09/12/2011    3010
17/02/2012    3007
              ... 
09/07/2010    2903
16/07/2010    2901
20/08/2010    2901
27/08/2010    2898
13/08/2010    2896
Name: count, Length: 143, dtype: int64


## Observations - Sales_DataSet

Nothing obvious is visual as far as extreme deviations

In [12]:
## Create Copy Of DataFrame
dfSales_DataSet_Work = dfSales_DataSet.copy()

## Strip Spaces From Date Column

In [13]:
#remove spaces from date column (string datatype)
for objColumn in["Date"]:
    dfSales_DataSet_Work[objColumn] = dfSales_DataSet_Work[objColumn].str.strip()
    
#show first 5 rows to check    
dfSales_DataSet_Work.head()    

,Store,Dept,Date,Weekly_Sales,IsHoliday
0,1,1,05/02/2010,24924.50,False
1,1,1,12/02/2010,46039.49,True
2,1,1,19/02/2010,41595.55,False
3,1,1,26/02/2010,19403.54,False
4,1,1,05/03/2010,21827.90,False


## Observations - Sales_DataSet

No data corruptions

## Validate Numerical Values

see if Stores column has values less than 1
see if Dept column has values less than 1
see if Weekly_Sales column have values less then 0

In [14]:
#get count of how many values in Stores column are less than 1
intLessThanOne = (dfSales_DataSet_Work["Store"] < 1).sum()
#print result
intLessThanOne

0

---

## Observations - Sales_DataSet

Values for Store in acceptable limits

## Validate Dept Values

In [15]:
#get count of how many values in Dept column are less than 1
intLessThanOne = (dfSales_DataSet_Work["Dept"] < 1).sum()
#print result
intLessThanOne

0

## Observations - Sales_DataSet

Values for Store in acceptable limits

## Validate Weekly_Sales Values

In [16]:
#Get count of how many values in Weekly_Sales column are less than 1
intLessThanOne = (dfSales_DataSet_Work["Weekly_Sales"] < 1).sum()
#print result
intLessThanOne

2006

## Observations - Sales_DataSet

Values for Weekly_Values include negative values
This suggests credits given, will leave data as is


## Data Transformations/Feature Engineering

Date - convert to datetime  

In [17]:
#convert Date to datetime
dfSales_DataSet_Work["Date"] = pd.to_datetime(dfSales_DataSet_Work["Date"], format="%d/%m/%Y")
#show results
dfSales_DataSet_Work.dtypes

Store                    int64
Dept                     int64
Date            datetime64[ns]
Weekly_Sales           float64
IsHoliday                 bool
dtype: object

In [18]:
#double check records
dfSales_DataSet_Work.head()

,Store,Dept,Date,Weekly_Sales,IsHoliday
0,1,1,2010-02-05,24924.50,False
1,1,1,2010-02-12,46039.49,True
2,1,1,2010-02-19,41595.55,False
3,1,1,2010-02-26,19403.54,False
4,1,1,2010-03-05,21827.90,False


## Observations - Sales_DataSet

Conversion successful (Strangely Displaying In American Date Foormat??)

## Remove Unnecessary Columns

In [19]:
#nothing to do as no viable columns to remove!


## Save Work DataFrame To CSV File

In [20]:
#create list of columns new and old JUST what is actually needed
lstColumns = [
                #original columns
                "Store","Dept","Date","Weekly_Sales","IsHoliday",
               ]

#create new DataFrame for cleaned data
dfCleaned = dfSales_DataSet_Work[lstColumns].copy()

#save dfSales_DataSet_Work with chatGPT suggested options
dfCleaned.to_csv(strCurrentDir + CNST_STR_CLEANFILEPATH,index=False, date_format="%d/%m/%Y")

#Note: excluded column: Dept as not in the hypothesis analysis which is a nice way
#      avoiding the fact I don't know just how to plot a chart with 98 departments
#      and have it readable!

# Conclusions and Next Steps

ETL went well, DataSet ready for use